### PR TITLE
Move Database Introspection to Classes

### PR DESCRIPTION
**Pull request by vincentsarago**
👋 @bitner 

This PR does:
- Update table_index SQL to `==` the one in `tifeatures`
- adapt for the new Table class
- add optional orjson and default JSON encoder/decoder

**breaking change**
- update `table_catalog` type to `Dict[str, Dict[str, Any]]`
- `Table` class changes
- rename `function_catalog` to `timvt_function_catalog` to avoid collision with tifeatures function_catalog

This reuse code from `tifeatures` (not yet public) project.

### CODE REVIEW

**Comment by vincentsarago** [comment]
test fails right now with `FAILED tests/routes/test_tiles.py::test_function_tile - asyncpg.exceptions.DataError: invalid input for query argument $6: {} (expected str, got bytes)` which I think is related to https://github.com/developmentseed/timvt/blob/828768bc133d1c3bb1c5e63b2929d6006abaade5/timvt/db.py#L19-L26 and https://github.com/developmentseed/timvt/blob/828768bc133d1c3bb1c5e63b2929d6006abaade5/timvt/layer.py#L275-L283

**Comment by vincentsarago** [comment]
## This PR
```
Run siege --file .github/workflows/data/urls_epsg4326.txt -b -c 20 -r 100 > /dev/null
** SIEGE 4.0.4
** Preparing 20 concurrent users for battle.
The server is now under siege...
Transactions:		        2000 hits
Availability:		      100.00 %
Elapsed time:		        4.37 secs
Data transferred:	        0.19 MB
Response time:		        0.04 secs
Transaction rate:	      457.67 trans/sec
Throughput:		        0.04 MB/sec
Concurrency:		       19.31
Successful transactions:           0
Failed transactions:	           0
Longest transaction:	        0.13
Shortest transaction:	        0.00
```

## Master 
```
Run siege --file .github/workflows/data/urls_epsg4326.txt -b -c 20 -r 100 > /dev/null
** SIEGE 4.0.4
** Preparing 20 concurrent users for battle.
The server is now under siege...
Transactions:		        2000 hits
Availability:		      100.00 %
Elapsed time:		        2.78 secs
Data transferred:	        0.19 MB
Response time:		        0.03 secs
Transaction rate:	      719.42 trans/sec
Throughput:		        0.07 MB/sec
Concurrency:		       19.38
Successful transactions:           0
Failed transactions:	           0
Longest transaction:	        0.07
Shortest transaction:	        0.00
``` 
😬

**Comment by vincentsarago** [comment]
FYI: There is a small impact on performance with this PR, we should wait before merging!

## This PR 
```
$ siege --file .github/workflows/data/urls_epsg4326.txt -b -c 20 -r 100 > /dev/null
** SIEGE 4.1.3
** Preparing 20 concurrent users for battle.
The server is now under siege...

Transactions:                   2000 hits
Availability:                 100.00 %
Elapsed time:                   3.03 secs
Data transferred:               5.47 MB
Response time:                  0.03 secs
Transaction rate:             660.07 trans/sec
Throughput:                     1.81 MB/sec
Concurrency:                   18.11
Successful transactions:        2000
Failed transactions:               0
Longest transaction:            0.08
Shortest transaction:           0.02
```

## Master
```
$ siege --file .github/workflows/data/urls_epsg4326.txt -b -c 20 -r 100 > /dev/null
** SIEGE 4.1.3
** Preparing 20 concurrent users for battle.
The server is now under siege...

Transactions:                   2000 hits
Availability:                 100.00 %
Elapsed time:                   2.73 secs
Data transferred:               5.47 MB
Response time:                  0.02 secs
Transaction rate:             732.60 trans/sec
Throughput:                     2.00 MB/sec
Concurrency:                   16.76
Successful transactions:        2000
Failed transactions:               0
Longest transaction:            0.09
Shortest transaction:           0.00
```

File: timvt/dbmodel.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+            cols = [c for c in cols if c in properties]
+
+        if len(cols) < 1:
+            raise TypeError("No columns selected")
+
+        return cols
+
+
+Database = Dict[str, Dict[str, Any]]
```
Comment (vincentsarago) [inline timvt/dbmodel.py:84]: `table_catalog` will now be of type: `Dict[str, Dict[str, Any]]`

File: timvt/dependencies.py [hunk 2]
```diff
@@ -64,9 +64,8 @@ def LayerParams(
         assert table_pattern.groupdict()["schema"]
         assert table_pattern.groupdict()["table"]
 
-        table_catalog = getattr(request.app.state, "table_catalog", [])
-        for r in table_catalog:
-            if r["id"] == layer:
-                return Table(**r)
+        table_catalog = getattr(request.app.state, "table_catalog", {})
+        if layer in table_catalog:
```
Comment (vincentsarago) [inline timvt/dependencies.py:68]: we now have a `dict` so we can check if the `layer` is in the dict

File: timvt/dbmodel.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+
+async def get_table_index(
+    db_pool: asyncpg.BuildPgPool,
+    schemas: Optional[List[str]] = ["public"],
```
Comment (vincentsarago) [inline timvt/dbmodel.py:89]: TODO: allow customization of `schemas` from settings

**Comment by vincentsarago** [comment]
Update on performance, I've ran more benchmark locally and the difference is really small and still faster than pg_tileserv so I think we are 👍

File: timvt/layer.py [hunk 4]
```diff
@@ -70,21 +71,24 @@ class Table(Layer):
...
-    geometry_srid: int
-    properties: Dict[str, str]
+
+    @root_validator
+    def bounds_default(cls, values):
+        """Get default bounds from the first geometry columns."""
+        geoms = values.get("geometry_columns")
+        if geoms:
+            values["bounds"] = geoms[0].bounds
+        return values
```
Comment (vincentsarago) [inline timvt/layer.py:91]: by default we get the layer bounds from the first geometry_columns cc @bitner

File: timvt/layer.py [hunk 5]
```diff
@@ -113,18 +117,16 @@ async def get_tile(
             "buffer", str(tile_settings.tile_buffer)
         )  # Size of extra data to add for a tile.
 
-        # create list of columns to return
-        geometry_column = self.geometry_column
-        geometry_srid = self.geometry_srid
-        cols = self.properties
-        if geometry_column in cols:
-            del cols[geometry_column]
+        geometry_column = self.geometry_column(kwargs.get("geom"))
```
Comment (vincentsarago) [inline timvt/layer.py:120]: we enable `geometry_column` selection with the `geom` query parameter

File: timvt/layer.py [hunk 6]
```diff
@@ -268,7 +270,7 @@ async def get_tile(
                     ":xmax",
                     ":ymax",
                     ":epsg",
-                    ":query_params",
+                    ":query_params::text::json",
```
Comment (vincentsarago) [inline timvt/layer.py:273]: had to add this to resolve the issue with the new json encoder/decoder!

**Review by bitner — APPROVED** [review]
Woo hoo! Thanks @vincentsarago

**Pull request merged**